File: ios/TrueImageResources.m
#import "TrueImageResources.h"

@implementation TrueImageResources

+ (UIImage *)imageNamed:(NSString *)name traits:(UITraitCollection *)traits
{
  return [UIImage imageNamed:name inBundle:NSBundle.mainBundle compatibleWithTraitCollection:traits];
}

+ (CGImageRef)rasterize:(UIImage *)image
                   size:(CGSize)size
                  scale:(CGFloat)scale
                   tint:(UIColor *)tint
                   mode:(TrueImageFitMode)mode
                 traits:(UITraitCollection *)traits
              pixelSize:(CGSize *)pixelSize
{
  if (size.width <= 0 || size.height <= 0) {
    return NULL;
  }
  UIGraphicsImageRendererFormat *format = [UIGraphicsImageRendererFormat new];
  format.scale = scale;
  format.opaque = NO;
  UIGraphicsImageRenderer *renderer = [[UIGraphicsImageRenderer alloc] initWithSize:size format:format];
  CGRect rect = TrueImageFitRect(CGRectMake(0, 0, size.width, size.height), image.size, mode);
  // Resolve dynamic colours against the current traits so a DynamicColorIOS
  // tint re-rasterises correctly on appearance changes.
  UIImage *drawn = image;
  if (tint) {
    drawn = [image imageWithTintColor:[tint resolvedColorWithTraitCollection:traits]
                        renderingMode:UIImageRenderingModeAlwaysOriginal];
  }
  UIImage *raster = [renderer imageWithActions:^(UIGraphicsImageRendererContext *_) {
    [drawn drawInRect:rect];
  }];
  if (pixelSize) {
    *pixelSize = CGSizeMake(image.size.width * image.scale, image.size.height * image.scale);
  }
  return CGImageRetain(raster.CGImage);
}

@end

File: ios/TrueImageResources.swift
import UIKit

/// Scheme-less sources name an asset catalog image. They are rasterised at
/// the view's size on the main thread, tinted, and drawn in the same frame.
enum TrueImageResources {
  struct Raster {
    let image: CGImage
    /// Intrinsic size of the catalog image in pixels, reported through onLoad.
    let pixelSize: CGSize
  }

  static func image(named name: String, traits: UITraitCollection) -> UIImage? {
    UIImage(named: name, in: .main, compatibleWith: traits)
  }

  static func raster(
    _ image: UIImage,
    size: CGSize,
    scale: CGFloat,
    tint: UIColor?,
    mode: TrueImageFitMode,
    traits: UITraitCollection
  ) -> Raster? {
    guard size.width > 0, size.height > 0 else { return nil }
    let format = UIGraphicsImageRendererFormat()
    format.scale = scale
    format.opaque = false
    let renderer = UIGraphicsImageRenderer(size: size, format: format)
    let rect = TrueImagePolicy.fitRect(
      bounds: CGRect(origin: .zero, size: size),
      intrinsic: image.size,
      mode: mode
    )
    // Resolve dynamic colours against the current traits so a
    // DynamicColorIOS tint re-rasterises correctly on appearance changes.
    let resolvedTint = tint?.resolvedColor(with: traits)
    let drawn = resolvedTint.map { image.withTintColor($0, renderingMode: .alwaysOriginal) } ?? image
    guard let cg = renderer.image({ _ in drawn.draw(in: rect) }).cgImage else { return nil }
    return Raster(
      image: cg,
      pixelSize: CGSize(width: image.size.width * image.scale, height: image.size.height * image.scale)
    )
  }
}
